### DISCUSSION TITLE
Make geosearch ranking rule more tolerant on distances

### DISCUSSION THREAD

**Discussion by ManyTheFish** [meilisearch/product#264] [Feedback & Feature Proposal]
## Summary

Today the geo-search ranking rule sort strictly the result by distance, meanings that the probability to have more than 1 document by buckets is null.
The issue to have a ranking rule that returns the document one by one is that all the ranking rules that follow the geo-search will apply their sort on one document, making them completely useless.

## Possible improvement

The geo-search ranking rule should sort by geo-points accepting a rate difference between the distance of the first geo-point found for the bucket and the distance of the other points.

For example; we take a rate difference of `50%` (arbitrary chosen):
- if the first point found is at a distance of 100 meters, all the points that have a distance under 150 meters will be considered as equivalent to the point at a distance of 100 meters.
- if the first point found is at a distance of 4,0 Km, all the points that have a distance under 6,0 Km will be considered as equivalent to the point at a distance of 4,0 Km.

**Comment by curquiza** [comment]
Related issues:
- https://github.com/meilisearch/MeiliSearch/issues/2081

**Comment by irevoire** [comment]
I would like to implement this.
But we already [have one user](https://github.com/meilisearch/MeiliSearch/issues/1872) with a use case where he needs to sort **exactly** its documents; I think we should propose an option with a good default to configure that at search time.

We could introduce a third optional parameter in the `_geoPoint` sort rule that would be a percentage like many said, for example.
So by default, the setting would be set to 50% (we'll see once it's implemented if it's a good default), and then our user could write `_geoPoint(52.3214, 41.6542, 0)` if he doesn't want to make any documents equals.

[Algolia also has a similar parameter](https://www.algolia.com/doc/guides/managing-results/refine-results/geolocation/in-depth/geo-ranking-precision/) but you need to set it for your whole query if I understood correctly, which is, in my opinion, a quite bad design.

I don't think [elastic-search](https://www.elastic.co/guide/en/elasticsearch/reference/7.3/search-request-body.html#geo-sorting) has a similar parameter.

What do you think of that? I think it's [quite close to what typesense did](https://typesense.org/docs/0.21.0/api/documents.html#sorting-by-additional-attributes-within-a-radius) so maybe it's not the best design, but I don't see any more user-friendly option than that :thinking: 

EDIT: Also I noticed that algolia & typesense only use this parameter with distances, meters or kilometres but no percentage.
So maybe @ManyTheFish we should reconsider your proposition or implement the two possibilities.
Something like that: 
- `_geoPoint(52.3214, 41.6542, 200m)`
- `_geoPoint(52.3214, 41.6542, 2km)`
- `_geoPoint(52.3214, 41.6542, 50%)`

**Reply by gmourier** [reply]
I really like the idea of the two notations. 

Indeed Algolia has separated most of the parameters of the GeoSearch at the root level. 

We preferred to centralize them around the concepts of filtering and sorting with built-in rules. It seems to me so much easier to have these rules boosted with parameters. We should stick to it for consistency.

**Reply by gmourier** [reply]
If the user specify 100% I guess the `_geoPoint` rule becomes "useless" and will have no precedence over other criteria? Am I right @irevoire?

**Reply by bidoubiwa** [reply]
Because root level parameters do not provide the possibility (at least with that design) to create multiple rules for multiple geo filters I agree that it is a bad idea.

I think we should as much as possible align to what others do in order to ease any migration from X to MeiliSearch. So about the distances, I think we should first align with what others do, thus, at the very least the user should be able to use 100m or 1km. We could, as a bonus, also give the possibility to users to define it with percentages. In this case, we would have to accept 3% as well and recognize the format based on the last symbols I guess.

**Reply by irevoire** [reply]
> If the user specify 100% I guess the `_geoPoint` rule becomes "useless" and will have no precedence over other criteria? Am I right @irevoire?

No, if we take the example @ManyTheFish wrote before let's look at what happens:
> For example; we take a rate difference of 50% (arbitrary chosen):
>
>    if the first point found is at a distance of 100 meters, all the points that have a distance under 150 meters will be considered as equivalent to the point at a distance of 100 meters.
>    if the first point found is at a distance of 4,0 Km, all the points that have a distance under 6,0 Km will be considered as equivalent to the point at a distance of 4,0 Km.

If we take a rate difference of 100%:
- The first point is found at a distance of 100 meters, then all the points that have a distance under 200 meters (= 100m + 100% of 100m) will be considered as equivalent.
- The first point is found at a distance of 4.0Km, all the point under 8Km will be considered the same
